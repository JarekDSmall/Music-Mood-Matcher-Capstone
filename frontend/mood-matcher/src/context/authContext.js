import React, { createContext, useContext, useState } from 'react';

export const AuthContext = createContext();

export const useAuth = () => {
    return useContext(AuthContext);
};

export const AuthProvider = ({ children }) => {
    const [currentUser, setCurrentUser] = useState(null);
    const [isAuthenticated, setIsAuthenticated] = useState(false);

    const login = (user) => {
        setCurrentUser(user);
        setIsAuthenticated(true);
    };

    const logout = () => {
        setCurrentUser(null);
        setIsAuthenticated(false);
    };

    const register = async (email, password) => {
        // Here, you'll typically call your backend API to register the user.
        // For now, I'll just simulate a successful registration.
        setCurrentUser({ email }); // This is just a mock. Replace with actual user data from your backend.
        setIsAuthenticated(true);
    };

    const contextValue = {
        currentUser,
        isAuthenticated,
        login,
        logout,
        register  // <-- Added the register function here
    };

    return (
        <AuthContext.Provider value={contextValue}>
            {children}
        </AuthContext.Provider>
    );
};

