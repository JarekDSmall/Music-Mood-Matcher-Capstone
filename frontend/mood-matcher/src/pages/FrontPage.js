import React, { useEffect, useState } from 'react';
import axios from 'axios';

function Login() {
    const [spotifyAuthUrl, setSpotifyAuthUrl] = useState('');

    useEffect(() => {
        const fetchSpotifyAuthUrl = async () => {
            try {
                const response = await axios.get('/spotify/auth-url');
                setSpotifyAuthUrl(response.data.authUrl);
            } catch (err) {
                console.error("Failed to fetch Spotify auth URL:", err);
            }
        };

        fetchSpotifyAuthUrl();
    }, []);

    return (
        <div className="login-container">
            <h2>Login</h2>
            {spotifyAuthUrl && (
                <a href={spotifyAuthUrl} className="spotify-login-button">
                    Login with Spotify
                </a>
            )}
        </div>
    );
}

class HomePage extends React.Component {
    render() {
        return (
            <div className="home-page">
                <header>
                    <h1>Welcome to Music Mood Matcher!</h1>
                </header>
                <main>
                    <p>
                        Discover music that matches your mood. Log in with Spotify to get started.
                    </p>
                    {/* You can add more components or content here */}
                </main>
                <footer>
                    <p>Music Mood Matcher &copy; 2023</p>
                </footer>
            </div>
        );
    }
}

function FrontPage() {
    return (
        <div>
            <HomePage />
            <Login />
        </div>
    );
}

export default FrontPage;
