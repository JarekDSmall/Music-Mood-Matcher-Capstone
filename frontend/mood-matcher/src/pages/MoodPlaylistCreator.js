import React, { useState, useEffect } from 'react';
import { searchTracks, createPlaylist, fetchUserProfile, fetchRecommendations, addTracksToPlaylist } from '../utility/spotifyAPI';


function MoodPlaylistCreator() {
    const [mood, setMood] = useState('');
    const [moodIntensity, setMoodIntensity] = useState(50);
    const [tracks, setTracks] = useState([]);
    const [userId, setUserId] = useState(null);

    useEffect(() => {
        const fetchUserId = async () => {
            try {
                const userProfile = await fetchUserProfile();
                setUserId(userProfile.id);
            } catch (error) {
                console.error("Error fetching user profile:", error);
            }
        };
        fetchUserId();
    }, []);

    const handleMoodChange = (event) => {
        setMood(event.target.value);
    };

    const handleIntensityChange = (event) => {
        setMoodIntensity(event.target.value);
    };

    const fetchTracks = async () => {
        try {
            // Search tracks based on mood to get seed tracks
            const seedTracks = await searchTracks(mood);
            const seedTrackIds = seedTracks.slice(0, 5).map(track => track.id); // Use the first 5 tracks as seeds

            // Fetch recommendations based on seed tracks
            const recommendedTracks = await fetchRecommendations(seedTrackIds);

            setTracks(recommendedTracks);
        } catch (error) {
            console.error("Error fetching tracks:", error);
        }
    };

    const handleSubmit = async () => {
        if (!userId) {
            console.error("User ID not available.");
            return;
        }
        try {
            const playlist = await createPlaylist(userId, `Mood Playlist - ${mood}`, `A playlist based on ${mood} mood with ${moodIntensity}% intensity.`);
            if (playlist && playlist.id) {
                // Add tracks to the created playlist
                const trackUris = tracks.map(track => track.uri);
                const success = await addTracksToPlaylist(playlist.id, trackUris);
                if (success) {
                    console.log("Tracks added to the playlist successfully.");
                    // Provide a link to the created playlist
                    window.alert(`Playlist created successfully! You can view it here: https://open.spotify.com/playlist/${playlist.id}`);
                } else {
                    console.error("Error adding tracks to the playlist.");
                }
            } else {
                console.error("Error creating playlist.");
            }
        } catch (error) {
            console.error("Error in handleSubmit:", error);
        }
    };
    

    return (
        <div>
            <h2>Create a Mood Playlist</h2>
            <form onSubmit={handleSubmit}>
                <div>
                    <label>Select Mood:</label>
                    <select value={mood} onChange={handleMoodChange}>
                        <option value="">--Select Mood--</option>
                        <option value="Happy">Happy</option>
                        <option value="Sad">Sad</option>
                        <option value="Energetic">Energetic</option>
                        <option value="Calm">Calm</option>
                        <option value="Relaxed">Relaxed</option>
                        <option value="Motivated">Motivated</option>
                        <option value="Angry">Angry</option>
                        <option value="Romantic">Romantic</option>
                        <option value="Melancholic">Melancholic</option>
                    </select>
                    {mood && (
                        <div>
                            <label>{`Intensity of ${mood}: ${moodIntensity}%`}</label>
                            <input 
                                type="range" 
                                min="0" 
                                max="100" 
                                value={moodIntensity} 
                                onChange={handleIntensityChange} 
                            />
                        </div>
                    )}
                </div>
                <button type="button" onClick={fetchTracks}>Fetch Tracks</button>
                <div>
                    <ul>
                        {tracks.map(track => (
                            <li key={track.id}>{track.name}</li>
                        ))}
                    </ul>
                </div>
                <button type="submit">Create Playlist</button>
            </form>
        </div>
    );
}

export default MoodPlaylistCreator;
