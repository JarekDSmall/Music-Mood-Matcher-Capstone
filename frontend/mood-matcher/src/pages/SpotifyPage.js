import React, { useState, useEffect } from 'react';
import axios from 'axios';
import jwtDecode from 'jwt-decode';

function SpotifyPage() {
    const token = localStorage.getItem('jwtToken');
    const decodedToken = jwtDecode(token);
    
    // Extract user-specific data from the decoded token
    const userId = decodedToken.id;
    const userName = decodedToken.name;
    const userEmail = decodedToken.email;

    const [isAuthenticated, setIsAuthenticated] = useState(false);
    const [topTracks, setTopTracks] = useState([]);

    useEffect(() => {
        const token = localStorage.getItem('spotifyAuthToken');
        if (token) {
            setIsAuthenticated(true);
            fetchTopTracks(token);
        }
    }, []);

    const fetchTopTracks = async (token) => {
        try {
            const response = await axios.get('/spotify/top-tracks', {
                headers: {
                    'Authorization': `Bearer ${token}`
                }
            });
            setTopTracks(response.data);
        } catch (error) {
            console.error("Error fetching top tracks:", error);
        }
    };

    const handleSpotifyLogin = () => {
        window.location.href = '/spotify-auth';
    };

    const handleSpotifyLogout = () => {
        localStorage.removeItem('spotifyAuthToken');
        setIsAuthenticated(false);
    };

    return (
        <div>
            <h1>Welcome, {userName}!</h1>
            {!isAuthenticated ? (
                <button onClick={handleSpotifyLogin}>Login with Spotify</button>
            ) : (
                <div>
                    <h2>Your Top Tracks</h2>
                    <ul>
                        {topTracks.map(track => (
                            <li key={track.id}>{track.name} by {track.artists[0].name}</li>
                        ))}
                    </ul>
                    <button onClick={handleSpotifyLogout}>Logout from Spotify</button>
                </div>
            )}
        </div>
    );
}

export default SpotifyPage;
