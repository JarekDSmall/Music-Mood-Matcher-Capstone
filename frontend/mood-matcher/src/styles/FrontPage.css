.home-page {
    background-color: #FFDB58; /* Mustard for a warm feel */
    color: #6B8E23; /* Olive Drab for text */
    padding: 20px;
    transition: background-color 0.3s ease;
}

.home-page header,
.home-page main,
.home-page footer {
    margin-bottom: 20px;
    transition: margin-bottom 0.3s ease;
}

/* Add any additional styles or hover effects you want for elements within .home-page */
