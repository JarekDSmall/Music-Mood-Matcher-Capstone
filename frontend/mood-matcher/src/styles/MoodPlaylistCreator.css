.mood-playlist-creator-container {
    background-color: #FFD700; /* Gold for a vibrant feel */
    color: #2F4F4F; /* Dark Slate Gray for text */
    padding: 20px;
    transition: background-color 0.3s ease;
}

.mood-playlist-creator-form {
    margin-bottom: 20px;
    transition: margin-bottom 0.3s ease;
}

.mood-playlist-creator-form label,
.mood-playlist-creator-form select,
.mood-playlist-creator-form input[type="range"] {
    margin-right: 10px;
    transition: margin-right 0.3s ease;
}

.mood-playlist-creator-form button {
    background-color: #DE3163; /* Cerise */
    color: #FFFFFF; /* White */
    padding: 10px 20px;
    border: none;
    border-radius: 5px;
    cursor: pointer;
    transition: background-color 0.3s ease, transform 0.3s ease;
}

.mood-playlist-creator-form button:hover {
    background-color: #C21807; /* Dark red */
    transform: scale(1.05); /* Slightly enlarge the button on hover */
}

.mood-playlist-creator-tracks-list {
    list-style-type: none;
    padding: 0;
}

.mood-playlist-creator-tracks-list li {
    background-color: #FF7F50; /* Coral */
    color: #FFFFFF; /* White */
    padding: 10px;
    margin-bottom: 5px;
    border-radius: 5px;
    transition: background-color 0.3s ease, transform 0.3s ease, box-shadow 0.3s ease;
}

.mood-playlist-creator-tracks-list li:hover {
    background-color: #FF6347; /* Tomato */
    transform: translateY(-3px); /* Slightly raise the item */
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.3); /* Shadow for a "lifted" effect */
}
