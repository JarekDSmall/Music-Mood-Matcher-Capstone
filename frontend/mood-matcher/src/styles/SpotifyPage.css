/* SpotifyPage.css */
.spotify-container {
    background-color: #F2D7D5; /* Light Pink for a soft background */
    color: #6B8E23; /* Olive Drab for text */
    padding: 20px;
    min-height: 100vh; /* Full height */
}

.spotify-header {
    color: #FF7F50; /* Coral for headers */
    margin-bottom: 20px;
}

.spotify-profile-picture {
    width: 150px;
    height: 150px;
    border-radius: 50%;
    margin-bottom: 10px;
    transition: transform 0.3s ease;
}

.spotify-profile-picture:hover {
    transform: scale(1.1);
}

.spotify-list {
    list-style-type: none;
    padding: 0;
}

.spotify-list-item {
    background-color: #FFD700; /* Gold for list items */
    color: #2F4F4F; /* Dark Slate Gray for list item text */
    padding: 10px;
    margin-bottom: 5px;
    border-radius: 5px;
    transition: background-color 0.3s ease;
}

.spotify-list-item:hover {
    background-color: #f0f0f0; /* Light grey for hover */
}

.spotify-button {
    background-color: #DE3163; /* Cerise for buttons */
    color: #FFFFFF; /* White for button text */
    padding: 10px 20px;
    border: none;
    border-radius: 5px;
    cursor: pointer;
    margin-right: 10px;
    transition: background-color 0.3s ease, transform 0.3s ease;
}

.spotify-button:hover {
    background-color: #C21807; /* Dark red for hover state */
    transform: scale(1.05);
}

/* Additional styles for logout button */
.logout-button {
    background-color: #e63946; /* Red for logout button */
    color: white;
    transition: background-color 0.3s ease;
}

.logout-button:hover {
    background-color: #d62828; /* Darker red for hover */
}

/* Smooth scrolling for the entire page */
html {
    scroll-behavior: smooth;
}

/* Fade in for list items */
@keyframes fadeIn {
    to {
        opacity: 1;
    }
}

.spotify-list-item {
    opacity: 0;
    animation: fadeIn 1s ease forwards;
}
