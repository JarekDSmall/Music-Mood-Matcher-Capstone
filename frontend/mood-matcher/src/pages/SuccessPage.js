import React from 'react';
import { useLocation, Link, useNavigate } from 'react-router-dom';
import { useAuth } from '../context/authContext'; // Adjust the path as necessary
import '../styles/SuccessPage.css';

function SuccessPage() {
    const location = useLocation();
    const navigate = useNavigate();
    const { logoutFromSpotify } = useAuth(); // Use logoutFromSpotify from your AuthContext
    const playlistId = location.state.playlistId;

    const handleFullLogout = async () => {
        try {
            logoutFromSpotify(); // Use logoutFromSpotify to log out
        } catch (error) {
            console.error("Failed to log out:", error);
        }
    };

    return (
        <div className="success-page-container">
            <h2 className="success-page-header">Playlist created successfully!</h2>
            <p>You can view it here: <a href={`https://open.spotify.com/playlist/${playlistId}`} target="_blank" rel="noopener noreferrer" className="success-page-link">Open Playlist</a></p>
            <p><Link to="/spotify" className="back-to-dashboard-link">Back to Dashboard</Link></p>
            <button onClick={handleFullLogout} className="success-page-button">Logout from Spotify</button>
        </div>
    );
}

export default SuccessPage;
