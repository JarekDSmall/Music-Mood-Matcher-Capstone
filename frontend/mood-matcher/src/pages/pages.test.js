import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter as Router } from 'react-router-dom';
import { useAuth } from '../context/authContext';
import FrontPage from './FrontPage';
import MoodPlaylistCreator from './MoodPlaylistCreator';
import SpotifyPage from './SpotifyPage';
import SuccessPage from './SuccessPage';
import renderer from 'react-test-renderer';

jest.mock('../context/authContext', () => ({
  useAuth: jest.fn(),
}));

describe('FrontPage Component', () => {
    beforeEach(() => {
        useAuth.mockImplementation(() => ({
            isAuthenticated: true, // Mock the value as per your test requirement
        }));
    });

    it('renders the main landing page', () => {
        render(
            <Router>
                <FrontPage />
            </Router>
        );
        expect(screen.getByText('Welcome to Music Mood Matcher!')).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const tree = renderer.create(
            <Router>
                <FrontPage />
            </Router>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
    // Add more tests for interactions or other functionalities
});

describe('MoodPlaylistCreator Component', () => {
    beforeEach(() => {
        useAuth.mockImplementation(() => ({
            // Mock any necessary auth context values or functions
        }));
    });

    it('renders mood selection', () => {
        render(
            <Router>
                <MoodPlaylistCreator />
            </Router>
        );
        expect(screen.getByText('Create a Mood Playlist')).toBeInTheDocument();
    });

    it('fetches tracks based on mood', () => {
        render(
            <Router>
                <MoodPlaylistCreator />
            </Router>
        );
        
        // Using getByText to select the dropdown option directly
        const moodOption = screen.getByText('Happy');
        fireEvent.click(moodOption);

        const fetchTracksButton = screen.getByText('Fetch Tracks');
        fireEvent.click(fetchTracksButton);
        // Add assertions for expected behavior
    });

    it('matches snapshot', () => {
        const tree = renderer.create(
            <Router>
                <MoodPlaylistCreator />
            </Router>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    // Add more tests for playlist creation and other functionalities
});

describe('SpotifyPage Component', () => {
    beforeEach(() => {
        useAuth.mockImplementation(() => ({
            // Mock any necessary auth context values or functions
        }));
    });

    it('renders Spotify login', () => {
        render(
            <Router>
                <SpotifyPage />
            </Router>
        );
        expect(screen.getByText('Login with Spotify')).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const tree = renderer.create(
            <Router>
                <SpotifyPage />
            </Router>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
    // Add tests for Spotify data fetching and other functionalities
});

describe('SuccessPage Component', () => {
    beforeEach(() => {
        useAuth.mockImplementation(() => ({
            logoutFromSpotify: jest.fn(), // Mock function or other values as needed
        }));
    });

    it('renders success message', () => {
        // Mock the location.state object with a playlistId
        const mockLocationState = {
            state: {
                playlistId: '12345' // Provide a mock playlistId
            }
        };

        render(
            <Router initialEntries={[mockLocationState]}>
                <SuccessPage />
            </Router>
        );

        expect(screen.getByText('Playlist created successfully!')).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const mockLocationState = {
            state: {
                playlistId: '12345'
            }
        };

        const tree = renderer.create(
            <Router initialEntries={[mockLocationState]}>
                <SuccessPage />
            </Router>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    // Add more tests for interactions or other functionalities
});

afterEach(() => {
    jest.clearAllMocks();
});
