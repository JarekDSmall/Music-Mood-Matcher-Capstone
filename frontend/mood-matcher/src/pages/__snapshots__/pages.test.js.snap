// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FrontPage Component matches snapshot 1`] = `
<div
  className="front-page-container"
>
  <div
    className="home-page"
  >
    <header
      className="home-header"
    >
      <h1>
        Welcome to Music Mood Matcher!
      </h1>
    </header>
    <main
      className="home-main"
    >
      <p>
        Discover music that matches your mood. Log in with Spotify to get started.
      </p>
      <h3>
        What is Music Mood Matcher?
      </h3>
      <p>
        Music Mood Matcher is a unique tool that curates playlists based on your current mood. Whether you're feeling happy, sad, energetic, or relaxed, we've got the perfect tracks to accompany your emotions.
      </p>
      <h3>
        How to use:
      </h3>
      <ol
        className="home-instructions"
      >
        <li>
          Login with your Spotify account.
        </li>
        <li>
          Select your current mood from the provided options.
        </li>
        <li>
          Adjust the intensity of your mood using the slider.
        </li>
        <li>
          Hit "Fetch Tracks" to get the list of songs that match your mood choice.
        </li>
        <li>
          Click "Create Playlist"
        </li>
        <li>
          The tracks will be curated in to a playlist called Mood Playlist depending on your mood! 
        </li>
        <li>
          Once created you can click the link to open the playlit on Spotify!
        </li>
      </ol>
    </main>
    <footer
      className="home-footer"
    >
      <p>
        Music Mood Matcher Â© 2023
      </p>
    </footer>
  </div>
   
</div>
`;

exports[`MoodPlaylistCreator Component matches snapshot 1`] = `
<div
  className="mood-playlist-creator-container"
>
  <h2
    className="mood-playlist-creator-title"
  >
    Create a Mood Playlist
  </h2>
  <form
    className="mood-playlist-creator-form"
    onSubmit={[Function]}
  >
    <div
      className="mood-selection-container"
    >
      <label
        className="mood-selection-label"
      >
        Select Mood:
      </label>
      <select
        className="mood-selection-select"
        onChange={[Function]}
        value=""
      >
        <option
          value=""
        >
          --Select Mood--
        </option>
        <option
          value="Happy"
        >
          Happy
        </option>
        <option
          value="Sad"
        >
          Sad
        </option>
        <option
          value="Energetic"
        >
          Energetic
        </option>
        <option
          value="Calm"
        >
          Calm
        </option>
        <option
          value="Relaxed"
        >
          Relaxed
        </option>
        <option
          value="Motivated"
        >
          Motivated
        </option>
        <option
          value="Angry"
        >
          Angry
        </option>
        <option
          value="Romantic"
        >
          Romantic
        </option>
        <option
          value="Melancholic"
        >
          Melancholic
        </option>
        <option
          value="Chill"
        >
          Chill
        </option>
        <option
          value="Uplifting"
        >
          Uplifting
        </option>
        <option
          value="Hopeful"
        >
          Hopeful
        </option>
        <option
          value="Mellow"
        >
          Mellow
        </option>
        <option
          value="Intense"
        >
          Intense
        </option>
        <option
          value="Groovy"
        >
          Groovy
        </option>
        <option
          value="Dreamy"
        >
          Dreamy
        </option>
        <option
          value="Nostalgic"
        >
          Nostalgic
        </option>
        <option
          value="Excited"
        >
          Excited
        </option>
        <option
          value="Pensive"
        >
          Pensive
        </option>
        <option
          value="Empowered"
        >
          Empowered
        </option>
      </select>
    </div>
    <button
      className="fetch-tracks-button"
      onClick={[Function]}
      type="button"
    >
      Fetch Tracks
    </button>
    <div
      className="track-list-container"
    >
      <ul
        className="mood-playlist-creator-tracks-list"
      />
    </div>
    <button
      className="create-playlist-button"
      type="submit"
    >
      Create Playlist
    </button>
  </form>
</div>
`;

exports[`SpotifyPage Component matches snapshot 1`] = `
<div
  className="spotify-container"
>
  <h1
    className="spotify-header"
  >
    Welcome, 
    !
  </h1>
  <p>
    Country: 
  </p>
  <p>
    Followers: 
    0
  </p>
  <button
    className="spotify-button"
    onClick={[Function]}
  >
    Login with Spotify
  </button>
</div>
`;

exports[`SuccessPage Component matches snapshot 1`] = `
<div
  className="success-page-container"
>
  <h2
    className="success-page-header"
  >
    Playlist created successfully!
  </h2>
  <p>
    You can view it here: 
    <a
      className="success-page-link"
      href="https://open.spotify.com/playlist/12345"
      rel="noopener noreferrer"
      target="_blank"
    >
      Open Playlist
    </a>
  </p>
  <p>
    <a
      className="back-to-dashboard-link"
      href="/spotify"
      onClick={[Function]}
    >
      Back to Dashboard
    </a>
  </p>
  <button
    className="success-page-button"
    onClick={[Function]}
  >
    Logout from Spotify
  </button>
</div>
`;
